README.md
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

![Bolder Flight Systems Logo](img/logo-words_75.png) &nbsp; &nbsp; ![Arduino Logo](img/arduino_logo_75.png)

# Airdata
This library contains functions for computing indicated airspeed, equivalent airspeed, true airspeed, pressure altitude, density altitude, and estimating outside air temperature (OAT) and air density. This library is compatible with Arduino and with CMake build systems.
   * [License](LICENSE.md)
   * [Changelog](CHANGELOG.md)
   * [Contributing guide](CONTRIBUTING.md)

# Installation

## Arduino
Use the Arduino Library Manager to install this library or clone to your Arduino/libraries folder. Additionally, the [Bolder Flight Systems Units library](https://github.com/bolderflight/units) must be installed. This library is added as:

```C++
#include "airdata.h"
```

An example Arduino executable is located at *examples/arduino/airdata_example/airdata_example.ino*. Teensy 3.x, 4.x, and LC devices are used for testing under Arduino and this library should be compatible with other devices.

## CMake
CMake is used to build this library, which is exported as a library target called *airdata*. The header is added as:

```C++
#include "airdata.h"
```

The library can be also be compiled stand-alone using the CMake idiom of creating a *build* directory and then, from within that directory issuing:

```
cmake ..
make
```

This will build the library, an example executable called *airdata_example*, and an executable for testing using the Google Test framework, called *airdata_test*. The example executable source file is located at *examples/cmake/airdata_example.cc*.

# Namespace
This library is within the namespace *bfs*.

# Constants

Standard day sea level speed of sound (m/s), pressure (Pa), and temperature(C, K) are defined as constants *SEA_LEVEL_SPEED_OF_SOUND_MPS*, *SEA_LEVEL_PRESSURE_PA*, *SEA_LEVEL_TEMPERATURE_C*, and *SEA_LEVEL_TEMPERATURE_K*.

```C++
float a_mps = bfs::SEA_LEVEL_SPEED_OF_SOUND_MPS;
float p0_pa = bfs::SEA_LEVEL_PRESSURE_PA;
float t0_c = bfs::SEA_LEVEL_TEMPERATURE_C;
float t0_k = bfs::SEA_LEVEL_TEMPERATURE_K;
```

Environmental lapse rate (K/m or C/m) is defined as a constant *LAPSE_RATE_KPM*.

```C++
float l = bfs::LAPSE_RATE_KPM;
```

The gas constant (J/(kg mol)) and molecular mas (kg/mol) of air are defined as constants *GAS_CONSTANT_JPKGMOL* and *MOLECULAR_MASS_AIR_KGPMOL*.

```C++
float r = bfs::GAS_CONSTANT_JPKGMOL;
float m = bfs::MOLECULAR_MASS_AIR_KGPMOL;
```

# Functions

**float Ias_mps(const float p)** Returns indicated airspeed (m/s) given differential pressure (Pa)

```C++
float ias = bfs::Ias_mps(500.0f);
std::cout << "IAS (m/s): " << ias << std::endl;  // 28.5457
```

**float Eas_mps(const float dp, const float sp)** Returns equivalent airspeed (m/s) given differential pressure (Pa) and static pressure (Pa)

```C++
float eas = bfs::Eas_mps(500.0f, 90000.0f);
std::cout << "EAS (m/s): " << eas << std::endl;  // 28.5429
```

**float Tas_mps(const float eas, const float t)** Returns true airspeed (m/s) given equivalent airspeed (m/s) and temperature (C)

``` C++
float tas = bfs::Tas_mps(eas, 17);
std::cout << "TAS (m/s): " << tas << std::endl;  // 28.6418
```

**float PressureAltitude_m(const float p)** Returns pressure altitue (m) given static pressure (Pa)

``` C++
float pa = bfs::PressureAltitude_m(90000);
std::cout << "Pressure Altitude (m): " << pa << std::endl;  // 988.1
```

**float DensityAltitude_m(const float p, const float t)** Returns density altitude (m) given static pressure (Pa) and temperature (C)

``` C++
float da = bfs::DensityAltitude_m(90000, 10);
std::cout << "Density Altitude (m): " << da << std::endl;  // 1039.88
```

**float Oat_c(const float t, const float agl)** Returns estimated outside air temperature (C) as a function of surface temperature (C) and altitude above ground (m). Uses a standard lapse rate of 6.5 C per km.

``` C++
float oat = bfs::Oat_c(15, 400);
std::cout << "OAT (C): " << oat << std::endl;  // 12.4
```
 
**float AirDensity_kgpm3(const float p, const float t)** Returns air density (kg/m^3) given pressure (Pa) and temperature (C)

``` C++
float density = bfs::AirDensity_kgpm3(101325, 15);
std::cout << "Air Density (kg/m^3): " << density << std::endl;  // 1.225
```


CHANGELOG.md
# Changelog

## v4.0.0
- Switched from templates to floats. Float should give enough precision for these computations and this should improve readability and compile times.

## v3.1.0
- Moved airdata specific constants from units to this library

## v3.0.4
- Updated units to v4.0.3

## v3.0.3
- Updated CMake required version to 3.14 for FetchContent_MakeAvailable

## v3.0.2
- Updated to Units v4.0.2

## v3.0.1
- Removed CMake define, which is no longer used in the preprocessor or build system

## v3.0.0
- Updated interface to work with Arduino and CMake build systems

## v2.1.0
- Templated functions against the input type

## v2.0.0
- Updated namespace to *bfs*
- Baselined against units v3.1.0

## v1.0.3
- Updated CONTRIBUTING
- Baselined against units v2.0.0
- Added line length limits to CI/CD and updated code to match

## v1.0.2
- Updated CONTRIBUTING
- Baselined against global_defs v1.1.2
- Updated *fetch_content* from ssh to https

## v1.0.1
- Changed license to MIT
- Baselined against global_defs v1.1.1

## v1.0.0
- Initial baseline


examples/cmake/airdata_example.cc
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2021 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/


#include <iostream>
#include "airdata.h"

int main() {
  float ias = bfs::Ias_mps(500.0f);
  std::cout << "IAS (m/s): " << ias << std::endl;  // 28.5457
  float eas = bfs::Eas_mps(500.0f, 90000.0f);
  std::cout << "EAS (m/s): " << eas << std::endl;  // 28.5429
  float tas = bfs::Tas_mps(eas, 17.0f);
  std::cout << "TAS (m/s): " << tas << std::endl;  // 28.6418
  float pa = bfs::PressureAltitude_m(90000.0f);
  std::cout << "Pressure Altitude (m): " << pa << std::endl;  // 988.518
  float da = bfs::DensityAltitude_m(90000.0f, 10.0f);
  std::cout << "Density Altitude (m): " << da << std::endl;  // 1039.88
  float oat = bfs::Oat_c(15.0f, 400.0f);
  std::cout << "OAT (C): " << oat << std::endl;  // 12.4
  float density = bfs::AirDensity_kgpm3(101325.0f, 15.0f);
  std::cout << "Air Density (kg/m^3): " << density << std::endl;  // 1.225
  return 0;
}
